### PR TITLE
Add entrypoint_service for pebble entrypoint services

### PR DESCRIPTION
**Pull request by linostar**
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

**Comment by linostar** [comment]
Tagging @deusebio for review.

File: rockcraft/models/project.py
Comment (cjdcordeiro) [inline rockcraft/models/project.py:340]: The user will see this message so it's better to be a bit more verbose (e.g. "The provided entrypoint-service is not a valid Pebble service"). Also, it should be `entrypoint-service` instead of `entrypoint_service` cause that's the internal variable name, and the users are not exposed to that

File: tests/spread/general/entrypoint-service/rockcraft.yaml
Comment (cjdcordeiro) [inline tests/spread/general/entrypoint-service/rockcraft.yaml:18]:
```suggestion
```

Comment (cjdcordeiro) [reply tests/spread/general/entrypoint-service/rockcraft.yaml:18]: (no need for it in this test :) )

File: tests/spread/general/entrypoint-service/rockcraft.yaml
Comment (cjdcordeiro) [inline tests/spread/general/entrypoint-service/rockcraft.yaml:12]: Just for the sake of being consistent with Pebble's MO:
 - the service name should allude to an actual service/process/daemon (e.g. "nginx", "controller", etc.). In this case, something like `test-service` fits the purpose
 - the `command` should be more configurable so that we can actually test the image in `task.yaml`. so I would suggest something like `command: echo [ foo ]`

File: tests/spread/general/entrypoint-service/task.yaml
Comment (cjdcordeiro) [inline tests/spread/general/entrypoint-service/task.yaml:7]:
```suggestion
```

File: tests/spread/general/entrypoint-service/task.yaml
Comment (cjdcordeiro) [inline tests/spread/general/entrypoint-service/task.yaml:10]:
```suggestion
  sudo /snap/rockcraft/current/bin/skopeo --insecure-policy copy oci-archive:entrypoint-service-test_latest_amd64.rock docker-daemon:entrypoint-service-test:latest
```

File: tests/spread/general/entrypoint-service/task.yaml
Comment (cjdcordeiro) [inline tests/spread/general/entrypoint-service/task.yaml:13]: why the `-v`? 

What I think we should be doing here is:
`docker run -d entrypoint-service-test` and test that the logs say "foo"

then
`docker run -d entrypoint-service-test bar` and test that the logs say "bar"

p.s. you don't need `--rm` cause the container isn't shutting down on its own and you're deleting later on, below

File: tests/unit/test_project.py
Comment (cjdcordeiro) [inline tests/unit/test_project.py:268]: could you also please test:
 - that the warning message is emitted
 - that the project loads well without this attribute. I say this because `entrypoint-service` should actually be part of ROCKCRAFT_YAML. this fixture should contain all the allowed fields, for the unmarshalling tests
 - when the value is empty

**Review by cjdcordeiro — CHANGES_REQUESTED** [review]
Thanks for this :)

Could you please bring the PR back to WIP? we should wait for @deusebio to finish some other changes on his side before we are 100% sure we need this feature.

Left a few comments ;)

File: docs/reference/rockcraft.yaml.rst
Comment (cjdcordeiro) [inline docs/reference/rockcraft.yaml.rst:175]: please make it come right after `services`. It feels more fluid that way.

Comment (cjdcordeiro) [inline docs/reference/rockcraft.yaml.rst:175]: this should be a Warning block. See an example in https://github.com/canonical/rockcraft/pull/410/files#diff-d88372eb092fe5a91cf204e108451fe4e9f34927f627138837af6b9083c1fb3bL175

File: docs/reference/rockcraft.yaml.rst
Comment (cjdcordeiro) [inline docs/reference/rockcraft.yaml.rst:170]:
```suggestion
``["/bin/pebble", "enter", "--verbose"]`` with ``["--args", "<serviceName>"]``.
```

File: rockcraft/models/project.py
Comment (cjdcordeiro) [inline rockcraft/models/project.py:330]: you shouldn't need the `always=True`, cause you don't want to do anything unless the field is actually set, which means you can also drop the `if entrypoint_service` below, cause the validator should only run when the field is set

File: rockcraft/models/project.py
Comment (cjdcordeiro) [inline rockcraft/models/project.py:338]: how come the linters aren't complaining about the size of this line? :open_mouth:

File: rockcraft/models/project.py
Comment (cjdcordeiro) [inline rockcraft/models/project.py:590]:
```suggestion
```

Comment (cjdcordeiro) [reply rockcraft/models/project.py:590]: added by mistake I believe...

File: rockcraft/services/package.py
Comment (cjdcordeiro) [inline rockcraft/services/package.py:172]: I think this should be split. It's better to have pure functions that only do one job. It also makes it easier for unit testing. So, I propose:

 - 2 functions: `new_image.set_entrypoint(entrypoint_service)` and `new_image.set_cmd(entrypoint_service_command)` 
 - then this code block becomes:
```python
# you let the function handle the `None` value
new_image.set_entrypoint(project.entrypoint_service)

if project.entrypoint_service:
    # you let this new function handle the splitting of the command
    new_image.set_cmd(project.services[project.entrypoint_service].command)
```

File: rockcraft/oci.py
Comment (cjdcordeiro) [inline rockcraft/oci.py:371]: this is the only line that matters in this function. the rest should be moved to a new function. See my explanation below in https://github.com/canonical/rockcraft/pull/410/files#diff-cf13f6f606f30afa23cf9954776527355fc805a6d81f60f29ec6899c261d83e9R172

File: tests/spread/general/entrypoint-service/task.yaml
Comment (cjdcordeiro) [inline tests/spread/general/entrypoint-service/task.yaml:17]: you're missing the `docker inspect` test we discussed offline, i.e.:

`docker inspect entrypoint-service-test:latest -f '{{json .Config.Entrypoint}}' == '["/bin/pebble","enter","--verbose", "--args", "test-service"]'`

and

`docker inspect entrypoint-service-test:latest -f '{{json .Config.Cmd}}' == '["foo"]'`

File: tests/unit/test_oci.py
Comment (cjdcordeiro) [inline tests/unit/test_oci.py:901]: you also need the test for when the entrypoint-service is not None

File: tests/unit/test_project.py
Comment (cjdcordeiro) [inline tests/unit/test_project.py:260]: this is weird. this shouldn't pass. "" is not a valid service.

**Review by cjdcordeiro — CHANGES_REQUESTED** [review]
I think it's almost there :)

File: tests/unit/test_project.py
Comment (linostar) [reply tests/unit/test_project.py:260]: An empty entrypoint_service will be considered the same as None.
Do you want to change that behavior?

File: rockcraft/models/project.py
Comment (linostar) [reply rockcraft/models/project.py:338]: Maybe because it's a triple-quote string? Found a bypass 😎

File: tests/unit/test_project.py
Comment (cjdcordeiro) [reply tests/unit/test_project.py:260]: I see.

Well, I guess both approaches could be right.

My gut feeling is to be conservative and not create an interface we would regret in the future. So maybe it is safer to not allow empty values for `entrypoint-service`.

File: docs/reference/rockcraft.yaml.rst
Comment (tigarmo) [inline docs/reference/rockcraft.yaml.rst:157]: I think this description is missing a link between the field and the pebble service. Something like `The optional name of the Pebble service to serve as the OCI entrypoint. If set, this makes Rockcraft extend...`

File: rockcraft/models/project.py
Comment (tigarmo) [inline rockcraft/models/project.py:335]:
```suggestion
        """Verify that the entrypoint_service exists in the services dict."""
```

File: tests/unit/test_project.py
Comment (tigarmo) [reply tests/unit/test_project.py:260]: https://docs.pydantic.dev/latest/api/types/#pydantic.types.constr

File: tests/unit/test_project.py
Comment (tigarmo) [inline tests/unit/test_project.py:291]:
```suggestion
    assert project.entrypoint_service is None
```

**Review by tigarmo — CHANGES_REQUESTED** [review]
Thanks! Left some comments, but I generally agree with @cjdcordeiro about not accepting empty strings in entrypoint-service and about cleaning up the set_entrypoint() method a bit

File: rockcraft/oci.py
Comment (cjdcordeiro) [inline rockcraft/oci.py:382]: only noticing this now, which means **we need another test**.

I'm 90% this won't work for most cases. basically, you're assuming that a service command will always be something like `command: app [ args ]`, but in reality, it can be `command: app -v --fixed-option --arg foo [ --some extra-arg --etc ]`.

This means:
 1. we need unit tests for assessing more variations of the service command
 2. this block of code needs to be more generic.

For 2. I suggest doing something like:

```python
# up top
import shlex
#...
def set_cmd(self, command: Optional[str]) -> None:
    """Set the OCI image CMD."""
    emit.progress("Configuring CMD...")
    image_path = self.path / self.image_name
    cmd_params = ["--clear=config.cmd"]
    command_sh_args = shlex.split(command)
    try:
        opt_args = command_sh_args[command_sh_args.index("[")+1:command_sh_args.index("]")]
    except ValueError:
        emit.debug(f"The entrypoint-service command '{command}' has no default arguments. CMD won't be set."
        return
    cmd_params.extend(opt_args)
    _config_image(image_path, cmd_params)
```
(haven't tested)

P.S.: why is `shlex` needed? Well, you must consider the following (which btw, is yet another unit test we need for this): `command: app --arg "[ foo ]" [ --optional ]`. IOW, if passed as a string, the squared brackets have no effect in pebble, and are simply treated as any other argument for the command app.

File: rockcraft/services/package.py
Comment (cjdcordeiro) [inline rockcraft/services/package.py:167]:
```suggestion
    if project.services:
```

The 2nd condition is redundant because if that wasn't the case, an exception would've already been raised in the project validation at https://github.com/canonical/rockcraft/pull/410/files#diff-54f38854c7e2d7d0fb5ab8160a493b4cf8f6f7da496a149eeff2680ec7ae6dbbR343

File: rockcraft/services/package.py
Comment (cjdcordeiro) [inline rockcraft/services/package.py:170]:
```suggestion
```

This is not needed. If it doesn't exist, don't set it

**Review by cjdcordeiro — CHANGES_REQUESTED** [review]
Almost there :) just found a bug, so we'll need more tests (see my comments below).

Also, please keep a linear history for the commits, because it becomes hard to know what was already reviewed whenever there is a force-push.

File: rockcraft/services/package.py
Comment (linostar) [reply rockcraft/services/package.py:167]: Actually, I had to revert this requested change because the linters start complaining because of the following line if the change is made:
https://github.com/canonical/rockcraft/actions/runs/6983627927&source=gmail&ust=1700936191049000&usg=AOvVaw06UHcYXWgZrPqgXarPnzoQ

File: rockcraft/oci.py
Comment (linostar) [reply rockcraft/oci.py:378]: `command` can be None, and splitting None in shlex is deprecated.

File: rockcraft/oci.py
Comment (linostar) [reply rockcraft/oci.py:377]: it was moved to set_entrypoint. otherwise if we keep it here we will have to call set_cmd in all cases. we've disccused it before iirc and you opted to have clear=config.cmd back in set_entrypoint.

File: rockcraft/models/project.py
Comment (linostar) [reply rockcraft/models/project.py]: ProjectValidationError will apparently need an argument, so the error string will be duplicated here too.
Check: https://github.com/canonical/rockcraft/actions/runs/7004567791/job/19052578345

File: rockcraft/oci.py
Comment (cjdcordeiro) [reply rockcraft/oci.py:377]: yes it should be in `set_entrypoint`. But what I meant is that it should also be in this fn. I get that right now, `cmd` has already been cleared when we get to this fn, and thus this might feel redundant. It's just that this is enforced by design. So I'm afraid that if tomorrow we change the ROCK's design such that one can set the CMD without setting the entrypoint, then this would be wrong.

File: rockcraft/models/project.py
Comment (cjdcordeiro) [reply rockcraft/models/project.py:361]: hum interesting...ok leave it then :shrug:

File: rockcraft/models/project.py
Comment (cjdcordeiro) [reply rockcraft/models/project.py:356]: Noted. We need a different strategy then, cause `ValueError` is not the right exception in this case. Let's break it down:
 - if either `[` or `]` don't exit, this code won't get to `raise ValueError`, and instead raise that exception by itself;
 - if it gets to line 356 (`raise ValueError`), then it's because the indexed of either `[` or `]` are wrong
 - in both cases, we want to raise a ProjectValidationError

I'd say: either raise `IndexError` or `ProjectValidationError` with an explanatory message.

**Review by tigarmo — APPROVED** [review]

**Review by cjdcordeiro — APPROVED** [review]
noice! It LGTM :) thanks for your patience.

**Comment by cjdcordeiro** [comment]
@deusebio any updates on your side?

**Comment by tigarmo** [comment]
@cjdcordeiro let me know if you want me to merge this

**Comment by cjdcordeiro** [comment]
will do @tigarmo . we need to wait for @deusebio

**Comment by deusebio** [comment]
@cjdcordeiro product is double checking that the breaking changes we discussed in Riga are fine. Anyhow, we have a PR open [here](https://github.com/canonical/charmed-spark-rock/pull/57) that I hope (if there are no concern with the changes) will land shortly

**Review by deusebio — APPROVED** [review]
LGTM, the modifications are rather clear to me, although I don't know the code base enough to provide a more in-depth review. 

I just left a small point/suggestion for the repo. Totally up yo you whether it is worth pursuing.

File: rockcraft/oci.py
Comment (linostar) [reply rockcraft/oci.py]: I provided default to None for both set_entrypoint and set_cmd now.
As for `str | None`, switching to it would require carrying over the change in all similar places in the project, to keep the consistency. Therefore I am opting for keeping it as it is for the time being.

**Comment by cjdcordeiro** [comment]
> LGTM, the modifications are rather clear to me, although I don't know the code base enough to provide a more in-depth review.
> 
> I just left a small point/suggestion for the repo. Totally up yo you whether it is worth pursuing.

thanks @deusebio 

How are things looking on your side? Do you have an updated version of your ROCK with the new Pebble services?

File: rockcraft/oci.py
Comment (linostar) [reply rockcraft/oci.py]: @cjdcordeiro it's not the entrypoint that is None here, but the entrypoint_service. Or am I missing something?

Comment (deusebio) [reply rockcraft/oci.py]: I didn't want to create a swirl here. I just meant that I kinda liked the syntax:

```
.set_entrypoint()
```

vs

```
.set_entrypoint(None)
```

by using default arguments. If this is not compliant to some internal rules you have (e.g. never provide default arguments), feel free to skip this. It was just a suggestion. 

The refactoring "Optional[str]" => " str | None", I agree it is out-of-scope as it touches a lot of places. It was also just a suggestion for your future development, provided by an outsider.

Both points do not block approval (which I have already given) nor require a re-review from my part!

**Comment by deusebio** [comment]
@cjdcordeiro we have the PR I linked above. As part of the CI, we also create an artifact [here](https://github.com/canonical/charmed-spark-rock/actions/runs/7023408765?pr=57). 

I'm just waiting Rob to approve the PR once we have green light about removing jupyter.

File: rockcraft/oci.py
Comment (cjdcordeiro) [reply rockcraft/oci.py]: :man_facepalming: you're right...read too fast. It's fine for the `entrypoint-service` arg to default to None. Nevermind me :p

**Comment by cjdcordeiro** [comment]
@deusebio Cool. Approved it. Thanks

@tigarmo once https://github.com/canonical/charmed-spark-rock/pull/57 is in, this PR can also go in :)

**Comment by deusebio** [comment]
@cjdcordeiro The other PR is now merged! :partying_face:

**Pull request merged**
